### PR TITLE
Feature/permissions fix

### PR DESCRIPTION
**Pull request by neilgaietto**
Changes:
- Updated to Umbraco Core 7.3.3
- Updated code to use ActionsResolver
- Fixed bug where changing a users role was not granting new permissions
- Fixed bug where changing a users role was not clearing the permissions granted by the previous role
- Added option to exclude setting permissions that were already granted to via base roles
- Added option to replace child permissions
- Added event on new nodes to store the appropriate role permissions

### CODE REVIEW

File: UserGroupPermissions/Events/ContentEvents.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+        private void ContentService_Created(IContentService sender, NewEventArgs<IContent> e)
+        {
+            if (e.Parent == null || e.Entity == null || !e.Entity.IsNewEntity()) return;
+            //copy parent permissions to new content node
+            _userTypePermissionsService.CopyNodePermissions(e.Parent, e.Entity);
+        }
+
+        private void ContentService_Deleted(IContentService sender, DeleteEventArgs<IContent> e)
+        {
```
Comment (Nicholas-Westby) [inline UserGroupPermissions/Events/ContentEvents.cs:53]: Will clearing permissions for deleted content cause problems if that content is restored from the recycle bin in the content section?

**Comment by Nicholas-Westby** [comment]
Amazing! I will review this over the weekend and pull it in if all my testing passes.

File: UserGroupPermissions/Events/ContentEvents.cs (hunk 1 above)
Comment (neilgaietto) [inline UserGroupPermissions/Events/ContentEvents.cs:53]: From testing it seems that this event only fired when I deleted an item directly out of the recycle bin and not from the content tree, so I think it should be fine. I did just notice that it wasn't firing when you cleared the whole bin all at once.

Comment (Nicholas-Westby) [inline UserGroupPermissions/Events/ContentEvents.cs:53]: Perfect. The bulk-delete oddity should not be an issue, as my guess is this deletion code is really a matter of cleaning up data and so is not mission critical.

**Comment by neilgaietto** [comment]
The package will need rebuild if you merge. I didn't think to update that too.

**Comment by Nicholas-Westby** [comment]
Don't think I'll be able to pull this in tonight. I need to validate that everything works with SQL Server CE, and this is the error that shows up in my error log when I attempt to create a user:

```
 2015-12-06 19:45:38,290 [P972/D11/T12] ERROR umbraco.userTasks - Failed to create the user. Error from provider: ProviderError
System.Data.SqlServerCe.SqlCeException (0x80004005): There was an error parsing the query. [ Token line number = 1,Token line offset = 10,Token in error = from ]
   at System.Data.SqlServerCe.SqlCeCommand.ProcessResults(Int32 hr)
   at System.Data.SqlServerCe.SqlCeCommand.CompileQueryPlan()
   at System.Data.SqlServerCe.SqlCeCommand.ExecuteCommand(CommandBehavior behavior, String method, ResultSetOptions options)
   at System.Data.SqlServerCe.SqlCeCommand.ExecuteNonQuery()
   at StackExchange.Profiling.Data.ProfiledDbCommand.ExecuteNonQuery()
   at Umbraco.Core.Persistence.PetaPocoCommandExtensions.<>c__DisplayClass1.<ExecuteNonQueryWithRetry>b__0()
   at Umbraco.Core.Persistence.FaultHandling.RetryPolicy.ExecuteAction[TResult](Func`1 func)
   at Umbraco.Core.Persistence.Database.Execute(String sql, Object[] args)
   at UserGroupPermissions.Businesslogic.UserTypePermissionsService.DeleteUsersRolePermissions(Int32 userId) in c:\r\usergrouppermissions\UserGroupPermissions\Businesslogic\UserTypePermissionsService.cs:line 247
   at UserGroupPermissions.Events.UserSaved.User_Saving(IUserService service, SaveEventArgs`1 e) in c:\r\usergrouppermissions\UserGroupPermissions\Events\UserSaved.cs:line 62
   at Umbraco.Core.Services.UserService.CreateUserWithIdentity(String username, String email, String passwordValue, IUserType userType)
   at Umbraco.Web.Security.Providers.UmbracoMembershipProvider`2.PerformCreateUser(String memberTypeAlias, String username, String password, String email, String passwordQuestion, String passwordAnswer, Boolean isApproved, Object providerUserKey, MembershipCreateStatus& status)
   at Umbraco.Core.Security.UmbracoMembershipProviderBase.PerformCreateUser(String username, String password, String email, String passwordQuestion, String passwordAnswer, Boolean isApproved, Object providerUserKey, MembershipCreateStatus& status)
   at umbraco.userTasks.PerformSave()
```

This results in user creation failing (i.e., no user is created). I'll have to fix that and review further another day.

**Comment by neilgaietto** [comment]
The query I added to delete previous role permissions doesn't look like its working on SQL CE. I can look into alternative ways to run the query.
Can you edit pull requests or do we just want to close this and I will make a new one once the bug is fixed?

**Comment by neilgaietto** [comment]
SQL CE apparently doesn't allow joins in delete queries. [umbracoUser2NodePermission] does not have an identity column, so it makes rewriting this query a bit difficult. 
Not sure when I'll get to this bug, but I think its important enough that it needs fixed. Otherwise, role permissions do not get changed correctly.

**Comment by Nicholas-Westby** [comment]
Not sure. I've never tried to update a pull request before.

With regard to the DELETE with a JOIN, perhaps something like this may work: http://stackoverflow.com/questions/12525979/mssql-compact-delete-query-with-join

**Comment by neilgaietto** [comment]
It looks like since the pull request is on a branch, rather than a commit, that its taking my changes.
The problem with that is since the [umbracoUser2NodePermission] uses a compound key you cant do an "IN" query like that, since it has to compare multiple column values.

I think I will have to make a code change that selects all the records first and then runs a delete against the results.

**Comment by Nicholas-Westby** [comment]
You might be able to concatenate multiple fields and do an "IN" query based on that. Performance would be terrible though (full table scan), so it might be easier just to do the SELECT followed by the DELETE (as you suggested).

**Comment by neilgaietto** [comment]
I got it working by pulling the records then deleting them, but I don't like it because the Database.Delete function does not accept an array so they are deleted individually. I will probably next try your suggestion of concatenating multiple fields if this is too slow, but I don't know how much more performant it would be. I am using this on a pretty large set of data, so I will find out pretty soon.

**Comment by Nicholas-Westby** [comment]
@neilgaietto I suspect pulling the records then deleting them individually will be faster for larger data sets (assuming a small number of deletions). Thanks for looking into this.

**Comment by Nicholas-Westby** [comment]
@neilgaietto  FYI, I merged this in, did a bit of code cleanup, and created a release (1.3.0). Sorry it took so long.

By the way, I noticed you were working on another branch. If it's anything you think I should add to the main code base, let me know. I can either accept a pull request, or if you want me to I can take a look at the code changes you made and manually add them.

**Comment by neilgaietto** [comment]
@Nicholas-Westby I believe my changes did not support SQL Server CE. Were you fine with dropping support for CE?

My other feature branch adds a new permission to disable Media Uploads based on user roles. Is that something you'd be interested in? If so then I will make a pull request, otherwise I need to get a hotfix out for this branch. There was an issue with copying deleted node permissions that needs applied here.

**Comment by Nicholas-Westby** [comment]
@neilgaietto I tested a few things on SQL Server CE and it seemed to work.

The ability to disable media uploads based on a permission sounds useful. And would definitely like that hotfix. Two pull requests would be the ideal case, but if you want to include them both in the same pull request, that works for me.

**Comment by neilgaietto** [comment]
I think you're right. I forgot I updated this branch to support SQL CE. 
I'll try to find some time in the next few days to put a couple branches together with the feature and hotfix.

**Comment by Nicholas-Westby** [comment]
@neilgaietto Awesome, looking forward to it. I promise to merge it in faster this time :-)

**Comment by neilgaietto** [comment]
@Nicholas-Westby I just realized its been a year since I created these changes, haha.
The bug I mentioned earlier should not be affected by the code you merged. It was only with some code in my other branch. So you should be good with the codes current state.

In my other branch I had to convert a lot of the sql operations that were done via code to actual sql statements. This offers a huge performance improvement, but unfortunately does not support SQL CE.

I'd recommend dropping SQL CE support if you have the option. Using the plugin without the raw sql inserts/deletes just does not work on larger site, which I think is the more common use case.

**Comment by Nicholas-Westby** [comment]
@neilgaietto Any hints as to which SQL statements (or C# code) in particular were underperforming? Maybe I can take a look. Umbraco's PetaPoco functionality should make it possible to create reasonably performant queries. And I'd hate to drop support for other providers (Umbraco supports SQL Server, SQL Server CE, and MySql).

**Comment by neilgaietto** [comment]
@Nicholas-Westby 
The problem areas are the parts that copy permissions or update permissions to many nodes. 
The main two senarios are:
1. Adding multiple new role permissions to a node with several hundred child nodes
2. Adding a user to a role when the role contains permissions for 100's of nodes. This also includes changing a users role.

The main issue is how umbraco stores permissions. Each permission is stored individually for each node. So even small role permission changes can add up to hundreds of updates. It's not to big of a deal if you can do a bulk insert/update, but as far as I can tell, that's not possible with SQL CE.

Some of the areas in code:
- UserTypePermissionsService.CopyPermissionsForSingleUser
This grabs all role permissions that the user is in and copies it to memory. Then it does 1 row insert at a time.

So if the user is in a role that has Read/Save/Publish/Sort/Audit on 100 nodes, then it will do 500 individual sql statements.

Versus raw sql where you can do a "select into" and have it completed in a single statement.


The same issue is with
- UserTypePermissionsService.CopyPermissions
If you apply Read/Save/Publish/Sort/Audit permissions for 2 roles with 5 users each to a node with 100 children, then it would perform 5050 updates just in the permissions table alone. This is not an unusual case, think new articles.
Again, this could all be completed in a single sql statement. But because SQL CE doesn't support more advanced sql statements, it all has to be done procedurally in code.

Deleting permissions is another one.

Really the code works fine until you are having to modify permissions in a site with many roles, users and nodes. Which I feel is the most common case to use this.

I agree that its not something I particularly want to do, but I also think any site that uses SQL CE is going to be so small that they won't need this feature. When I need a site with roles, its usually a very large site with a lot of data, so you need to segment sections out. My site for instance has 6 roles, 40+ users, and several thousand nodes. It would take 15+ minutes to update role permissions on some sections of the site using the old method. Now its down to a fraction of that.

If you can figure it out using umbraco that would be great. Let me know if you can, because I couldn't find a way around this restriction.

**Comment by Nicholas-Westby** [comment]
Thanks for the info! I'll give it my best shot to try optimizing it. Since I'm working with a site I expect to have about 1,000+ users, 10+ roles, and hundreds or thousands of content nodes, it sounds like I'll need to address this one way or another.

**Comment by neilgaietto** [comment]
Here's the branch where I solved the issue with the new sql scripts, so you can see the difference:
https://github.com/neilgaietto/usergrouppermissions/blob/feature/Disable-Media-Upload-Permission/UserGroupPermissions/Businesslogic/UserTypePermissionsService.cs

1000+ umbraco users sounds pretty extreme. I really hope umbraco implements a better solution for their roles soon. Or at least a better way to override their permissions. If they opened up a hook to override permissions then we wouldn't have to deal with performance issues.

**Comment by Nicholas-Westby** [comment]
@neilgaietto Thanks for the code example. Here's my first attempt to optimize `CopyPermissionsForSingleUser`: https://github.com/Nicholas-Westby/usergrouppermissions/blob/f69e12760171903e0ed6c267e05173b74df8acae/UserGroupPermissions/Businesslogic/UserTypePermissionsService.cs#L128

The key bit is the call to `BulkInsertRecords`. For new SQL Server versions, that will perform a bulk insertions. For SQL Server CE, it will fallback to a bunch of SQL inserts. That way, you get good performance where it's possible without breaking compatibility with all database variants.

I only tested it once and it seemed to work, but I haven't confirmed the performance yet. My assumption is that all the LINQ I'm doing is getting compiled down to a single sensible SQL statement, but I'll have to do some SQL tracing to confirm that.

If that turns out to be correct, perhaps I can apply similar techniques to the other areas you've identified as needing performance improvements.

**Comment by Nicholas-Westby** [comment]
I used SQL Profiler to see what my above code was doing with a SQL Server database when I clicked "Apply Group Permissions" on a user:

trace

It's basically loading up 3 tables and performing a bulk insert. Here are some notes for each numbered item in the screenshot:

1. Load the entire `umbracoUser2NodePermission` table.
2. Load the entire `umbracoNode` table.
3. Load the entire `UserTypePermissions` table.
4. Perform a bulk insert into the `umbracoUser2NodePermission` table.

None of those particularly worry me, except for the first one. With thousands of users and thousands of nodes, that could potentially load hundreds of megabytes into memory.

I should be able to modify the first item so it only loads up the rows for a single user rather than for all users, which should mean it will remain under a megabyte of data (except in even more extreme scenarios, in which case it might creep up to multiple megabytes).

I should also be able to select only the columns I'm interested in from the umbracoNode table (the second item above) and I should be able to select only the rows for a single user type from the umbracoUser2NodePermission table (the third item above).

Once I've done all of the above, the performance should be much improved, at least for the call to `CopyPermissionsForSingleUser`.

**Comment by neilgaietto** [comment]
Nice work, it already is an improvement to what is there now.
First time noticing BulkInsertRecords, and your right, it does look like the best bet as to getting it working for all versions of SQL.

Instead of doing all those queries, cant you just write that into a single sql statement? The insert was the issue, not selecting. You can create a query like this and avoid all the in-memory joins. (This was freehand from my code so it might not run)

` var newPermissions = _sqlHelper.Fetch<newPermissionDTO>("SELECT @0, a.[NodeId], a.[PermissionId] FROM [UserTypePermissions] a " +
                        "JOIN [umbracoNode] n on a.NodeId = n.id " +
                        "LEFT OUTER JOIN[umbracoUser2NodePermission] b on a.NodeId = b.nodeId and a.PermissionId = b.permission and b.userId = @0 " +
                        "WHERE a.[UserTypeId] = @1 AND B.userId is null ", userId, userTypeId);
`


I'll me interested to see how it performance with a lot of data.

**Comment by Nicholas-Westby** [comment]
That was fairly painless. Here's the new trace:

trace

And this is essentially what each query is doing:

1. `SELECT userId, nodeId, permission FROM umbracoUser2NodePermission WHERE userId = 1`
2. `SELECT id FROM umbracoNode`
3. `SELECT * FROM UserTypePermissions WHERE userTypeId = 1`
4. Bulk insert.

Looks pretty performant.

**Comment by Nicholas-Westby** [comment]
@neilgaietto Good point. I probably could simplify the select queries. Though, I want to be sure I do it in a way that is compatible with all databases used by Umbraco (e.g., I'd want to confirm that all can do LEFT OUTER JOIN in the same way).

**Comment by Nicholas-Westby** [comment]
@neilgaietto  FYI, I made a few more changes in hopes it makes everything optimized and compatible with SQL Server CE. Haven't had a chance to fully test it, but thought I'd let you know in case you wanted to give it a go. Had to do some funky things to replace the DELETE's that had JOIN's (didn't seem to work with CE). Latest code is on the master branch.

At some point, I'll do some testing. I'm thinking of creating 1,000 nodes, 1,000 users, and 10 or so user types, then do a few different operations to see if it works relatively quickly on SQL Server.

**Comment by Nicholas-Westby** [comment]
Hmmm, looks like I've managed to mess up my own permissions as an admin (e.g., I can't delete content nodes anymore). Will have to look into that one.

**Comment by Nicholas-Westby** [comment]
@neilgaietto FYI, I fixed up the admin permission issues and created a new release (https://github.com/Nicholas-Westby/usergrouppermissions/releases/tag/v1.4.0). Seems to be very performant now. If you feel like giving it a test run, I'd much appreciate it. Thanks for all of your help thus far!

**Comment by neilgaietto** [comment]
Great, I will try and find some time over the weekend to test the performance with my site.

**Comment by Nicholas-Westby** [comment]
@neilgaietto Just curious, have you had a chance to give it a test run?

**Comment by neilgaietto** [comment]
I started yes, but haven't finished.  
Testing against my existing site was being troublesum so I want to redo it with a fresh site. I was having issues with duplicate permission records being inserted.  Not sure the cause yet but the site had a previous issue of sometimes only partially applying permission because the query would timeout. But yeah, I still want to finish testing this.

**Comment by Nicholas-Westby** [comment]
@neilgaietto Just checking in to see if you've discovered any issues. I may have a bit of time this weekend to fix any issues, so the sooner I know about them the better.

**Comment by neilgaietto** [comment]
I did a pass through of the changes but didn't get to the depth that I wanted. So no issues to report. I think the only question I had was if you could simplify the 3 queries in CopyPermissionsForSingleUser into a single query.

**Comment by Nicholas-Westby** [comment]
@neilgaietto Cool, I created issue #13, so I'll hopefully be able to look at that this weekend.

**Comment by Nicholas-Westby** [comment]
@neilgaietto OK, simplified the queries as part of this commit: https://github.com/Nicholas-Westby/usergrouppermissions/commit/d364ff8e1995fcaec9d87f483b67993ace76e604

Let me know if you notice anything else.

**Comment by Nicholas-Westby** [comment]
@neilgaietto Oh, forgot to mention one thing. I got rid of a join to the `umbracoNode` table. Didn't seem necessary.

**Comment by neilgaietto** [comment]
That looks much better. Did you get a chance to see if it performed any better? Not sure if I have time any time soon, but now that its all one query it should be quicker.
Did you need to use the query builder to support sql ce or did it offer other advantages? It looks fine, it just seems like it adds unnecessary code without making it any easier to follow.

**Comment by Nicholas-Westby** [comment]
@neilgaietto I'm not familiar with the SQL syntax in all SQL flavors supported by Umbraco, and I also want to use the provided abstractions in case they change something in the future (e.g., if they decide to support a new SQL flavor). I figure it's there for a reason, so I might as well use it.

Not sure of how well it performs compared to the prior implementation, but it only took a second or two to run on a test install with hundreds of content nodes and hundreds of users (though, the permissions were already applied, so it would probably be a bit slower when actually changing permissions).

**Pull request merged**